### DISCUSSION TITLE
2022-42: OpenDAL Key-Value Adapter

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#383] [Announcements]
# 2022-42: OpenDAL Key-Value Adapter

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/reports/2022-42/

**Comment by noirgif** [comment]
对一个界面（interface，或者说接口？）来说，人可以透明地削减它的表达能力（expressiveness，想不到什么好词），但不能透明地扩张它地表达能力。所以要透明地统一界面就会伴随着部分表达能力的牺牲。

我觉得，文件系统这个界面本质上是键值的一个子集：从路径到文件的抽象。所以我的理解是，openDAL为了统一访问界面，牺牲了一部分表达能力（如range query）而将一部分kv抽象成文件系统。

我们来看缓存。我认为对于后端多种多样的openDAL来说，完全透明的缓存是不实际的——人们需要在某种程度上控制缓存策略。也就是说，我们是在扩充其表达能力，扩充了本来的POSIX文件系统抽象，变成了某种程度上能控制缓存的POSIX文件系统。当然只是在界面角度而言，实际我没见过这种POSIX。

无论是增加缓存，还是 kv 抽象到 kv，和openDAL之前所做的事情（kv/文件系统抽象到文件系统）是不同的。所以在假设这里的kv抽象是必要的前提下，我觉得放在另外一个模块里实现也不失为一种选择。

**Reply by Xuanwo** [reply]
> 我觉得，文件系统这个界面本质上是键值的一个子集：从路径到文件的抽象。所以我的理解是，openDAL为了统一访问界面，牺牲了一部分表达能力（如range query）而将一部分kv抽象成文件系统。

这个感觉比较牵强，文件系统比单纯的 path -> file content 的抽象还是复杂很多的，包括但不限于 get metadata，readdir, pread, pwrite 等。OpenDAL 做的事情实际上是选取了一个自己定义的抽象：对文件系统来说，是放弃了很多已有的能力；对 kv 来说，是增加了一层新的抽象，其内容对用户已经不再透明了。

> 我们来看缓存。我认为对于后端多种多样的openDAL来说，完全透明的缓存是不实际的——人们需要在某种程度上控制缓存策略。也就是说，我们是在扩充其表达能力，扩充了本来的POSIX文件系统抽象，变成了某种程度上能控制缓存的POSIX文件系统。当然只是在界面角度而言，实际我没见过这种POSIX。

这个其实有点理解上的偏差。OpenDAL 自己本身不做提供控制缓存的能力，这是由应用来控制的。比如说用户可以初始化两个 operator，一个是 s3，一个是 redis，然后他自己来控制中间的缓存逻辑。OpenDAL 只是提供一个统一访问的接口而已。为了满足大部分简单的场景，OpenDAL 也在 layer 里面提供了 metadata 和 content cache 的中间件，用户可以自己选。
